examples/psbt.rs

```diff
@@ -0,0 +1,711 @@
+    let value = utxo.value;
+
+    let script = match psbt_input.witness_script {
+        Some(ref witness_script) => witness_script.clone(),
+        None => {
+            if utxo.script_pubkey.is_v0_p2wpkh() {
+                p2wpkh_script_code(&utxo.script_pubkey)
+            } else if psbt_input
+                .redeem_script
+                .as_ref()
+                .map(Script::is_v0_p2wpkh)
+                .unwrap_or(false)
+            {
+                p2wpkh_script_code(psbt_input.redeem_script.as_ref().unwrap())
+            } else {
+                return Err(SignError::MissingWitnessScript);
+            }
+        }
+    };
+
+    Ok((
+        SighashCache::new(&psbt.unsigned_tx).segwit_signature_hash(
+            input_index,
+            &script,
+            value,
+            sighash.ecdsa_hash_ty().map_err(|_| SignError::Ecdsa)?,
+        ).map_err(|_| SignError::Sighash)?,
+        sighash.ecdsa_hash_ty().map_err(|_| SignError::Ecdsa)?
+    ))
+}
+
+/// Finalizes a PSBT accord to the Input Finalizer role described in BIP174.
+fn finalize_psbt(mut psbt: Psbt) -> Psbt {
+    use bitcoin::util::psbt::serialize::Serialize;
+
+    // Input 0: legacy UTXO
+
+    let sigs: Vec<_> = psbt.inputs[0].partial_sigs.values().collect();
+    let script_sig = script::Builder::new()
+        // OP_CHECKMULTISIG bug pops +1 value when evaluating so push OP_0.
+        // TODO: Consider adding OP_ZERO alias to rust-bitcoin (`pub static OP_ZERO: All = all::OP_PUSHBYTES_0;`).
+        .push_opcode(opcodes::all::OP_PUSHBYTES_0)
+        .push_slice(&sigs[0].serialize())
+        .push_slice(&sigs[1].serialize())
+        .push_slice(&psbt.inputs[0].redeem_script.clone().unwrap().serialize())
+        .into_script();
+
+    psbt.inputs[0].final_script_sig = Some(script_sig);
+
+    psbt.inputs[0].partial_sigs = BTreeMap::new();
+    psbt.inputs[0].sighash_type = None;
+    psbt.inputs[0].redeem_script = None;
+    psbt.inputs[0].bip32_derivation = BTreeMap::new();
+
+    // Input 1: SegWit UTXO
+
+    let script_sig = script::Builder::new()
+        .push_slice(&psbt.inputs[1].redeem_script.clone().unwrap().serialize())
+        .into_script();
+
+    psbt.inputs[1].final_script_sig = Some(script_sig);
+
+    let sigs: Vec<_> = psbt.inputs[1].partial_sigs.values().collect();
+    let mut script_witness: Witness = Witness::new();
+    script_witness.push([]);    // Pushes 0x00 to the stack.
+    script_witness.push(&sigs[1].serialize());
+    script_witness.push(&sigs[0].serialize());
+    script_witness.push(&psbt.inputs[1].witness_script.clone().unwrap().serialize());
+
+    psbt.inputs[1].final_script_witness = Some(script_witness);
+
+    psbt.inputs[1].partial_sigs = BTreeMap::new();
+    psbt.inputs[1].sighash_type = None;
+    psbt.inputs[1].redeem_script = None;
+    psbt.inputs[1].witness_script = None;
+    psbt.inputs[1].bip32_derivation = BTreeMap::new();
+
+    psbt
+}
+
+fn p2wpkh_script_code(script: &Script) -> Script {
```

Comment:
I would like to know _why_ this code is needed and why its not in `rust-bitcoin`, said another way, why do we need to drop the first two bytes from the script?

Reply:
We drop the first two bytes of the scriptPubKey as`script` because the first is the version (0x00) the second length of the pubkeyhash  (0x20) and the value p2wpkh script code wants here is the witness program AKA the pubkeyhash in p2wpkh

src: this [libbitcoin doc](https://github.com/libbitcoin/libbitcoin-system/wiki/P2WPKH-Transactions#spending-a-p2wpkh-output) and [the bip 141 spec](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki).

Perhaps it should live in `rust-bitcoin` because it's de/serialization of standard script

Reply:
Legend, cheers.